util/encoding.go

```diff
@@ -0,0 +1,243 @@
+3.Repeat the previous steps, substituting the remainder for N, until a remainder of 0 is reached.
+4.Place an additional 1 after the rightmost digit in the code word.
+
+So a Fibonacci encoded bit sequence is all about marking the corresponding bit as 1 while leaving the others as 0.
+Here are some examples:
+
+	FibonacciInt(1) = 11			1-> (1*1)->1
+	FibonacciInt(2) = 011 			2-> (1*1+2*1)->01
+	FibonacciInt(3) = 0011			3-> (1*0+2*0+3*1)->001
+	FibonacciInt(4) = 1011			4-> (1*1+2*0+3*1)->101
+	FibonacciInt(12) = 101011		12->(1*1+2*0+3*1+5*0+8*1)->10101
+
+According to the algorithm, no sequential ones can be found in the bit sequence, and the final bit is always a single 1,
+which makes it easy to extract several continuous Fibonacci encoded integers out from a bit stream if we
+place an additional 1 in the end, since 11 could be an end-of-sequence indicator. The advantage of this is,
+that a sequence of numbers can be encoded into a sequence of bits without the need to know the length of the
+bit sequences in advance.
+*/
+func (bs *BitStream) WriteFibonacciInt(num uint16) error {
+	// The num should be [1,6765). Actually once the num is larger than or equal to 987,
+	// the efficiency of Fibonacci Encoding would be no better than 'WriteUint16'.
+	if num <= 0 || num >= fibonacci(fibLen) {
+		return fibEncodeNumOutOfRangeErr
+	}
+	// Binary Search to find the largest fibonacci number less than or equal to num.
+	lo, hi := 2, fibLen-1
+	for lo < hi {
+		mid := (lo + hi) / 2
+		if num >= fibonacci(mid) && num < fibonacci(mid+1) {
+			lo = mid
+			break
+		}
+		if num < fibonacci(mid) {
+			hi = mid - 1
+		} else {
+			lo = mid + 1
+		}
+	}
+	// Calculate the fibonacci-encoded sequence.
+	var fibEncoded uint32 = 1
+	offset := 1
+	for i := lo; i >= 2; i-- {
+		if num >= fibLookup[i] {
+			num -= fibLookup[i]
+			fibEncoded |= 1 << offset
+		}
+		offset++
+	}
+	encodedLength := lo
+	fibEncoded <<= 32 - encodedLength
+
+	bs.appendNBits([]byte{
+		byte(fibEncoded >> 24),
+		byte(fibEncoded >> 16),
+		byte(fibEncoded >> 8),
+		byte(fibEncoded),
+	}, uint16(encodedLength))
+	return nil
+}
+
+/*
+WriteIntRange writes an IntRange object into the BitStream.
+
+The basic idea is 12 bits for size, 1 bit indicating if a range refers to a single number or not, followed by
+one or two Fibonacci encoded numbers, and then followed by other ranges fell behind.
+
+For example, an IntRange consisting of [1, 1], [3, 4] would be encoded into 000000000010 0 11 1 0011 11.
+There are two ranges, so the size is encoded into 000000000010. The first range [1, 1] refers to a single integer 1,
+so the indicator is 0, followed by Fibonacci encoded 1, while the next range [3, 4] refers to a range containing
+not only one integer, so the indicator is 1, followed by Fibonacci encoded 3 and 1 (4-3).
+*/
+func (bs *BitStream) WriteIntRange(intRange *IntRange) error {
```

Comment:
Shouldn't Fibonacci enter the name someplace as the GPP spec also defines a non-Fibonacci int range. The non-Fibonacci range will be needed once TCFv2 support is added.

Reply:
Would you like to comment on the difference between the Fibonacci Int range and the legacy Int Range that is used in TCF. We do plan to add TCFv2 code to this library once we get some time to do it. It hasn't been too urgent as an independent TCFv2 library exists.

Reply:
The current name will be an issue when TCFv2 is brought into the project and the original Int range is implemented. The documentation refers to them as `Range (Fibonacci)` and `Range (Int)`, so the current name could confuse people looking for the `Range (Fibonacci)` implementation. This is why the corresponding read function is called `ReadFibonacciRange()`.